From 327f15ce9d5e2c783b06a8e0bfb03d61090d7f03 Mon Sep 17 00:00:00 2001
From: moneta <lorenzo.moneta@cern.ch>
Date: Mon, 19 Nov 2018 15:14:17 +0100
Subject: [PATCH] Fix ROOT-9801, by correctly copying the members for lazy
 initialization in TFormula::Copy. Add also a test for reproducing ROOT-9801

---
 hist/hist/src/TFormula.cxx  |  6 +++--
 test/TFormulaParsingTests.h | 48 +++++++++++++++++++++++++++++++++++++
 2 files changed, 52 insertions(+), 2 deletions(-)

diff --git a/hist/hist/src/TFormula.cxx b/hist/hist/src/TFormula.cxx
index 0bcbc7832a3..7dcb9f4be96 100644
--- a/hist/hist/src/TFormula.cxx
+++ b/hist/hist/src/TFormula.cxx
@@ -669,6 +669,8 @@ void TFormula::Copy(TObject &obj) const
    fnew.fClingInitialized = fClingInitialized;
    fnew.fAllParametersSetted = fAllParametersSetted;
    fnew.fClingName = fClingName;
+   fnew.fSavedInputFormula = fSavedInputFormula;
+   fnew.fLazyInitialization = fLazyInitialization;
 
    // case of function based on a C++  expression (lambda's) which is ready to be compiled
    if (fLambdaPtr && TestBit(TFormula::kLambda)) {
@@ -3421,8 +3423,8 @@ void TFormula::ReInitializeEvalMethod() {
       R__LOCKGUARD(gROOTMutex);
 
       // std::cout << "gClingFunctions list" << std::endl;
-      // for (auto thing : gClingFunctions)
-      //    std::cout << "gClingFunctions : " << thing.first << std::endl;
+      //  for (auto thing : gClingFunctions)
+      //     std::cout << "gClingFunctions : " << thing.first << std::endl;
 
       auto funcit = gClingFunctions.find(fSavedInputFormula);
 
